src/document/doc-utils.js

```diff
@@ -180,41 +180,123 @@ function removeLines(doc) {
+        parts.length -= 2;
+      }
+
+      if (parts.length > 0) {
+        const lastPart = stripDocTrailingHardlineFromDoc(
+          parts[parts.length - 1]
+        );
+        parts[parts.length - 1] = lastPart;
+      }
+      return { ...doc, parts };
+    }
+    case "align":
+    case "indent":
+    case "group":
+    case "line-suffix": {
+      const contents = stripDocTrailingHardlineFromDoc(doc.contents);
+      return { ...doc, contents };
+    }
+    case "if-break": {
+      const breakContents = stripDocTrailingHardlineFromDoc(doc.breakContents);
+      const flatContents = stripDocTrailingHardlineFromDoc(doc.flatContents);
+      return { ...doc, breakContents, flatContents };
+    }
   }
-  return parts;
+
+  return doc;
 }
 
-function stripTrailingHardline(doc, withInnerParts = false) {
+function stripTrailingHardline(doc) {
   // HACK remove ending hardline, original PR: #1984
-  if (doc.type === "concat" && doc.parts.length !== 0) {
-    const parts = withInnerParts ? getInnerParts(doc) : doc.parts;
-    const lastPart = parts[parts.length - 1];
-    if (lastPart.type === "concat") {
+  return stripDocTrailingHardlineFromDoc(cleanDoc(doc));
+}
+
+const isConcat = (doc) => doc && doc.type === "concat";
+function cleanDocFn(doc) {
+  switch (doc.type) {
+    case "fill":
+      if (doc.parts.length === 0 || doc.parts.every((part) => part === "")) {
+        return "";
+      }
+      break;
+    case "group":
+      if (!doc.contents && !doc.id && !doc.break && !doc.expandedStates) {
+        return "";
+      }
+      // Remove nested only group
       if (
-        lastPart.parts.length === 2 &&
-        lastPart.parts[0].hard &&
-        lastPart.parts[1].type === "break-parent"
+        doc.contents.type === "group" &&
+        doc.contents.id === doc.id &&
+        doc.contents.break === doc.break &&
+        doc.contents.expandedStates === doc.expandedStates
       ) {
-        return { type: "concat", parts: parts.slice(0, -1) };
+        return doc.contents;
       }
+      break;
+    case "align":
+    case "indent":
+    case "line-suffix":
+      if (!doc.contents) {
+        return "";
+      }
+      break;
+    case "if-break":
+      if (!doc.flatContents && !doc.breakContents) {
+        return "";
+      }
+      break;
+  }
 
-      return {
-        type: "concat",
-        parts: doc.parts.slice(0, -1).concat(stripTrailingHardline(lastPart)),
-      };
+  if (!isConcat(doc)) {
+    return doc;
+  }
+
+  const newParts = [];
+  for (const part of doc.parts) {
+    if (!part) {
+      continue;
+    }
+    const previousPart = newParts[newParts.length - 1];
+    const isPreviousPartString = typeof previousPart === "string";
+    const [currentPart, ...rest] = isConcat(part) ? part.parts : [part];
+    if (isPreviousPartString && typeof currentPart === "string") {
+      newParts[newParts.length - 1] += currentPart;
+    } else {
+      newParts.push(currentPart);
     }
+    newParts.push(...rest);
   }
```

Comment:
So, this is the same thing as `normalizeParts` now, I suggest we move your implementation to that function and call it here.

Reply:
Feels not safe for this part https://github.com/prettier/prettier/blob/e15e6112b617ecadaae76fc624378231146fb0e9/src/language-html/printer-html.js#L387-L395

Reply:
You're right, I didn't think about that. Let's don't do this then.

Reply:
Reverted.